service/storage_service.cc

```diff
@@ -3658,9 +3675,25 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
             do_remove_node(endpoint);
         }
     } else if (existing && *existing == endpoint) {
-        tmptr->del_replacing_endpoint(endpoint);
+        // This branch in taken by replace-with-different-ip.
+        // We actually don't remove anything from replacing endpoints here, since
+        // we pass new host_id to del_replacing_endpoint. The same was true
+        // with IP-based token_metadata - the new IP was passed to del_replacing_endpoint
+        // and it didn't remove anything. I'm not sure why we need this code at all.
+        // The old IP is removed by the next block of logic - we detach old IP from token ring,
+        // it gets added to candidates_for_removal, then storage_service::remove_endpoint ->
+        // _gossiper.remove_endpoint -> storage_service::on_remove -> it's removed
+        // from topology and from replacing_endpoints.
+
+        tmptr->del_replacing_endpoint(host_id);
     } else {
-        tmptr->del_replacing_endpoint(endpoint);
+        // This branch is taken by replace-with-same-ip.
```

Comment:
This I also don't understand.

In replace-with-same-IP, how can we take this branch?
Suppose we call `handle_state_normal(endpoint)` where `endpoint` is the replacing (new) node. (This means we are finishing the replace operation.)
This:
```cpp
    auto host_id = _gossiper.get_host_id(endpoint);
```
returns the host ID of the new node.

So this:
```cpp
auto existing = tmptr->get_endpoint_for_host_id_if_known(host_id);
```
returns nullopt then? Why isn't the replacing endpoint already present in tmptr? Wasn't it inserted to token_metadata somewhere earlier during the replace operation?

If it didn't return nullopt, it would be `*existing == endpoint` and we would end up in the previous above.

Reply:
I  thought you modified topology to store mapping for both old and new host ID for the duration of replace. Which would mean that `existing` is not nullopt

Reply:
> I checked this code manually by running all individual tests in topology_custom/test_replace and checking by the logs that the state of topology and replacing endpoints is correct.

These tests run in raft-topology mode, `handle_state_normal` is for gossip mode.
Unless you modified the configuration to disable raft-topology?

Reply:
I modified `scylla_cluster.py` by temporarily removing `consistent-topology-changes` and `tablets`.